`@Fallback` strategy seems not to work with Kotlin coroutines and additional exception in method signature

Hi.
I am trying to adjust my Kotlin code to use "coroutines" while still defining a `@Fallback` method:

```kotlin
    @Incoming("lorawan")
    @Acknowledgment(Acknowledgment.Strategy.MANUAL)
    @Asynchronous
    @Retry(delay = 25, maxRetries = 5)
    @ExponentialBackoff
    @Fallback(fallbackMethod = "fallbackProcessor")
    suspend fun processor(kafkaMessage: Message<LoRaWanMessage>) {
       ...
    }
```

The following fallback method signature works as expected:
```
    suspend fun fallbackProcessor(kafkaMessage: Message<LoRaWanMessage>) {
       ...
    }
```

However, if I add an `Exception` like documented in the [Fallback Methods with Exception Parameter
](https://smallrye.io/docs/smallrye-fault-tolerance/6.2.2/reference/fallback.html#fallback-methods-with-exception-parameter):
```
    suspend fun fallbackProcessor(kafkaMessage: Message<LoRaWanMessage>, ex: Exception) {
       ...
    }
```

I get a `java.lang.RuntimeException: java.lang.RuntimeException: Failed to start quarkus` error:
```
Caused by: jakarta.enterprise.inject.spi.DeploymentException: org.eclipse.microprofile.faulttolerance.exceptions.FaultToleranceDefinitionException: Invalid @Fallback on LoRaUplinkKafkaGateway.processor(org.eclipse.microprofile.reactive.messaging.Message, kotlin.coroutines.Continuation): can't find fallback method 'fallbackProcessor' with matching parameter types and return type
	at io.quarkus.smallrye.faulttolerance.runtime.SmallRyeFaultToleranceRecorder.createFaultToleranceOperation(SmallRyeFaultToleranceRecorder.java:44)
	at io.quarkus.deployment.steps.SmallRyeFaultToleranceProcessor$processFaultToleranceAnnotations1906679055.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.SmallRyeFaultToleranceProcessor$processFaultToleranceAnnotations1906679055.deploy(Unknown Source)
	... 68 more
Caused by: org.eclipse.microprofile.faulttolerance.exceptions.FaultToleranceDefinitionException: Invalid @Fallback on LoRaUplinkKafkaGateway.processor(org.eclipse.microprofile.reactive.messaging.Message, kotlin.coroutines.Continuation): can't find fallback method 'fallbackProcessor' with matching parameter types and return type
	at io.smallrye.faulttolerance.config.FallbackConfig.validate(FallbackConfig.java:42)
	at io.smallrye.faulttolerance.config.FaultToleranceOperation.validate(FaultToleranceOperation.java:365)
	at io.quarkus.smallrye.faulttolerance.runtime.SmallRyeFaultToleranceRecorder.createFaultToleranceOperation(SmallRyeFaultToleranceRecorder.java:28)
	... 70 more
```

I am again overseeing something obvious in the documentation?
`io.smallrye:smallrye-fault-tolerance-kotlin` is on the classpath.
I am using Quarkus version 3.0.3 and smallrye-fault-tolerance 6.2.2.

Without Kotlin coroutines the fallback method with additional Exception in the signature work as expected.

--- a/implementation/autoconfig/core/src/main/java/io/smallrye/faulttolerance/autoconfig/KotlinSupport.java
+++ b/implementation/autoconfig/core/src/main/java/io/smallrye/faulttolerance/autoconfig/KotlinSupport.java
@@ -2,9 +2,11 @@ package io.smallrye.faulttolerance.autoconfig;
 
 // TODO this would ideally live in the `kotlin` module
 final class KotlinSupport {
+    private static final String KOTLIN_CONTINUATION = "kotlin.coroutines.Continuation";
+
     static boolean isLegitimate(MethodDescriptor method) {
         if (method.parameterTypes.length > 0
-                && method.parameterTypes[method.parameterTypes.length - 1].getName().equals("kotlin.coroutines.Continuation")
+                && method.parameterTypes[method.parameterTypes.length - 1].getName().equals(KOTLIN_CONTINUATION)
                 && method.name.endsWith("$suspendImpl")) {
             return false;
         }

--- a/implementation/fault-tolerance/src/main/java/io/smallrye/faulttolerance/FaultToleranceInterceptor.java
+++ b/implementation/fault-tolerance/src/main/java/io/smallrye/faulttolerance/FaultToleranceInterceptor.java
@@ -69,8 +69,6 @@ import io.smallrye.faulttolerance.core.fallback.FallbackFunction;
 import io.smallrye.faulttolerance.core.invocation.AsyncSupport;
 import io.smallrye.faulttolerance.core.invocation.AsyncSupportRegistry;
 import io.smallrye.faulttolerance.core.invocation.Invoker;
-import io.smallrye.faulttolerance.core.invocation.NormalMethodInvoker;
-import io.smallrye.faulttolerance.core.invocation.SpecialMethodInvoker;
 import io.smallrye.faulttolerance.core.invocation.StrategyInvoker;
 import io.smallrye.faulttolerance.core.metrics.CompletionStageMetricsCollector;
 import io.smallrye.faulttolerance.core.metrics.MetricsCollector;
@@ -98,6 +96,7 @@ import io.smallrye.faulttolerance.core.util.DirectExecutor;
 import io.smallrye.faulttolerance.core.util.ExceptionDecision;
 import io.smallrye.faulttolerance.core.util.SetBasedExceptionDecision;
 import io.smallrye.faulttolerance.core.util.SetOfThrowables;
+import io.smallrye.faulttolerance.internal.FallbackMethod;
 import io.smallrye.faulttolerance.internal.FallbackMethodCandidates;
 import io.smallrye.faulttolerance.internal.InterceptionInvoker;
 import io.smallrye.faulttolerance.internal.InterceptionPoint;
@@ -116,9 +115,6 @@ import io.smallrye.faulttolerance.metrics.MetricsProvider;
 @FaultToleranceBinding
 @Priority(Interceptor.Priority.PLATFORM_AFTER + 10)
 public class FaultToleranceInterceptor {
-
-    private static final Object[] EMPTY_ARRAY = {};
-
     private final Bean<?> interceptedBean;
 
     private final FaultToleranceOperationProvider operationProvider;
@@ -550,28 +546,13 @@ public class FaultToleranceInterceptor {
 
         if (candidates != null) {
             fallbackFunction = ctx -> {
-                Method fallbackMethod = candidates.select(ctx.failure.getClass());
+                FallbackMethod fallbackMethod = candidates.select(ctx.failure.getClass());
                 if (fallbackMethod == null) {
                     throw sneakyThrow(ctx.failure);
                 }
 
-                boolean isDefault = fallbackMethod.isDefault();
-                InvocationContext interceptionContext = ctx.invocationContext.get(InvocationContext.class);
                 try {
-                    Object[] arguments = interceptionContext.getParameters();
-                    if (arguments == null) {
-                        arguments = EMPTY_ARRAY;
-                    }
-                    if (fallbackMethod.getParameterCount() == arguments.length + 1) {
-                        Object[] argumentsWithException = new Object[arguments.length + 1];
-                        System.arraycopy(arguments, 0, argumentsWithException, 0, arguments.length);
-                        argumentsWithException[argumentsWithException.length - 1] = ctx.failure;
-                        arguments = argumentsWithException;
-                    }
-
-                    Invoker invoker = isDefault
-                            ? new SpecialMethodInvoker(fallbackMethod, interceptionContext.getTarget(), arguments)
-                            : new NormalMethodInvoker(fallbackMethod, interceptionContext.getTarget(), arguments);
+                    Invoker<?> invoker = fallbackMethod.createInvoker(ctx);
                     return asyncSupport == null ? (V) invoker.proceed() : (V) asyncSupport.toCompletionStage(invoker);
                 } catch (InvocationTargetException e) {
                     throw sneakyThrow(e.getCause());

--- a/implementation/fault-tolerance/src/main/java/io/smallrye/faulttolerance/config/KotlinSupport.java
+++ b/implementation/fault-tolerance/src/main/java/io/smallrye/faulttolerance/config/KotlinSupport.java
@@ -7,9 +7,11 @@ import java.lang.reflect.WildcardType;
 
 // TODO this would ideally live in the `kotlin` module
 final class KotlinSupport {
+    private static final String KOTLIN_CONTINUATION = "kotlin.coroutines.Continuation";
+
     static boolean isSuspendingFunction(Method method) {
         int params = method.getParameterCount();
-        return params > 0 && method.getParameterTypes()[params - 1].getName().equals("kotlin.coroutines.Continuation");
+        return params > 0 && method.getParameterTypes()[params - 1].getName().equals(KOTLIN_CONTINUATION);
     }
 
     static Type getSuspendingFunctionResultType(Method method) {

--- a/implementation/fault-tolerance/src/main/java/io/smallrye/faulttolerance/internal/FallbackMethod.java
+++ b/implementation/fault-tolerance/src/main/java/io/smallrye/faulttolerance/internal/FallbackMethod.java
@@ -1,0 +1,72 @@
+package io.smallrye.faulttolerance.internal;
+
+import java.lang.reflect.Method;
+import java.util.Arrays;
+
+import jakarta.interceptor.InvocationContext;
+
+import io.smallrye.faulttolerance.core.fallback.FallbackContext;
+import io.smallrye.faulttolerance.core.invocation.Invoker;
+import io.smallrye.faulttolerance.core.invocation.NormalMethodInvoker;
+import io.smallrye.faulttolerance.core.invocation.SpecialMethodInvoker;
+
+public final class FallbackMethod {
+    private static final Object[] EMPTY_ARRAY = {};
+
+    private final int exceptionParameterPosition; // < 0 if no exception parameter exists
+    private final Method method;
+
+    static FallbackMethod withoutExceptionParameter(Method method) {
+        return method == null ? null : new FallbackMethod(method, -1);
+    }
+
+    static FallbackMethod withExceptionParameter(Method method, int exceptionParameterPosition) {
+        return method == null ? null : new FallbackMethod(method, exceptionParameterPosition);
+    }
+
+    private FallbackMethod(Method method, int exceptionParameterPosition) {
+        this.method = method;
+        this.exceptionParameterPosition = exceptionParameterPosition;
+    }
+
+    // ---
+
+    public Invoker<?> createInvoker(FallbackContext<?> ctx) throws ReflectiveOperationException {
+        InvocationContext interceptionContext = ctx.invocationContext.get(InvocationContext.class);
+        Object[] arguments = interceptionContext.getParameters();
+        if (arguments == null) {
+            arguments = EMPTY_ARRAY;
+        }
+        arguments = adjustArguments(arguments, ctx.failure);
+
+        return method.isDefault()
+                ? new SpecialMethodInvoker<>(method, interceptionContext.getTarget(), arguments)
+                : new NormalMethodInvoker<>(method, interceptionContext.getTarget(), arguments);
+    }
+
+    private Object[] adjustArguments(Object[] arguments, Throwable exception) {
+        if (method.getParameterCount() == arguments.length) {
+            return arguments;
+        }
+
+        if (method.getParameterCount() == arguments.length + 1) {
+            Object[] argumentsWithException = new Object[arguments.length + 1];
+
+            int exceptionParameterPosition = this.exceptionParameterPosition;
+            for (int i = 0; i < arguments.length + 1; i++) {
+                if (i < exceptionParameterPosition) {
+                    argumentsWithException[i] = arguments[i];
+                } else if (i == exceptionParameterPosition) {
+                    argumentsWithException[i] = exception;
+                } else { // i > exceptionParameterPosition
+                    argumentsWithException[i] = arguments[i - 1];
+                }
+            }
+
+            return argumentsWithException;
+        }
+
+        throw new IllegalArgumentException("Cannot adjust arguments " + Arrays.toString(arguments)
+                + " to fallback method " + method);
+    }
+}

--- a/implementation/fault-tolerance/src/main/java/io/smallrye/faulttolerance/internal/FallbackMethodCandidates.java
+++ b/implementation/fault-tolerance/src/main/java/io/smallrye/faulttolerance/internal/FallbackMethodCandidates.java
@@ -10,15 +10,21 @@ import java.util.Set;
 import org.eclipse.microprofile.faulttolerance.exceptions.FaultToleranceException;
 
 public final class FallbackMethodCandidates {
-    private final Method withoutExceptionParam;
-    private final Map<Class<?>, Method> withExceptionParam;
+    private final FallbackMethod withoutExceptionParam;
+    private final Map<Class<?>, FallbackMethod> withExceptionParam;
 
     private FallbackMethodCandidates(Method withoutExceptionParam, Set<Method> withExceptionParam) {
-        this.withoutExceptionParam = withoutExceptionParam;
+        this.withoutExceptionParam = FallbackMethod.withoutExceptionParameter(withoutExceptionParam);
 
-        Map<Class<?>, Method> map = new HashMap<>();
+        Map<Class<?>, FallbackMethod> map = new HashMap<>();
         for (Method method : withExceptionParam) {
-            map.put(method.getParameterTypes()[method.getParameterCount() - 1], method);
+            int exceptionParameterPosition = method.getParameterCount() - 1;
+            if (KotlinSupport.isSuspendingFunction(method)) {
+                exceptionParameterPosition--;
+            }
+
+            map.put(method.getParameterTypes()[exceptionParameterPosition],
+                    FallbackMethod.withExceptionParameter(method, exceptionParameterPosition));
         }
         this.withExceptionParam = map;
     }
@@ -27,11 +33,11 @@ public final class FallbackMethodCandidates {
         return withoutExceptionParam == null && withExceptionParam.isEmpty();
     }
 
-    public Method select(Class<? extends Throwable> exceptionType) {
+    public FallbackMethod select(Class<? extends Throwable> exceptionType) {
         if (!withExceptionParam.isEmpty()) {
             Class<?> type = exceptionType;
             while (type != null) {
-                Method candidate = withExceptionParam.get(type);
+                FallbackMethod candidate = withExceptionParam.get(type);
                 if (candidate != null) {
                     return candidate;
                 }

--- a/implementation/fault-tolerance/src/main/java/io/smallrye/faulttolerance/internal/KotlinSupport.java
+++ b/implementation/fault-tolerance/src/main/java/io/smallrye/faulttolerance/internal/KotlinSupport.java
@@ -1,0 +1,25 @@
+package io.smallrye.faulttolerance.internal;
+
+import java.lang.reflect.Method;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+
+// TODO this would ideally live in the `kotlin` module
+final class KotlinSupport {
+    private static final String KOTLIN_CONTINUATION = "kotlin.coroutines.Continuation";
+
+    static boolean isSuspendingFunction(Method method) {
+        int params = method.getParameterCount();
+        return params > 0 && method.getParameterTypes()[params - 1].getName().equals(KOTLIN_CONTINUATION);
+    }
+
+    static boolean isSuspendingFunction(Type[] parameterTypes) {
+        int params = parameterTypes.length;
+        if (params > 0) {
+            Type last = parameterTypes[params - 1];
+            return last instanceof Class && last.getTypeName().equals(KOTLIN_CONTINUATION)
+                    || last instanceof ParameterizedType && last.getTypeName().startsWith(KOTLIN_CONTINUATION);
+        }
+        return false;
+    }
+}

--- a/implementation/fault-tolerance/src/main/java/io/smallrye/faulttolerance/internal/SecurityActions.java
+++ b/implementation/fault-tolerance/src/main/java/io/smallrye/faulttolerance/internal/SecurityActions.java
@@ -27,6 +27,7 @@ import java.security.AccessController;
 import java.security.PrivilegedAction;
 import java.security.PrivilegedActionException;
 import java.security.PrivilegedExceptionAction;
+import java.util.Arrays;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
@@ -125,8 +126,8 @@ final class SecurityActions {
 
         Class<?> clazz = beanClass;
         while (true) {
-            Set<Method> methods = getMethodsFromClass(declaringClass, clazz, name, expectedParameterTypes,
-                    expectedReturnType, expectedExceptionParameter, actualMapping, expectedMapping);
+            Set<Method> methods = getMethodsFromClass(clazz, name, expectedParameterTypes, expectedReturnType,
+                    expectedExceptionParameter, declaringClass, actualMapping, expectedMapping);
             for (Method method : methods) {
                 if (possibleFallbackMethodNames.contains(method.getName())) {
                     result.add(method);
@@ -145,8 +146,8 @@ final class SecurityActions {
         }
 
         for (Class<?> iface : beanClass.getInterfaces()) {
-            Set<Method> methods = getMethodsFromClass(declaringClass, iface, name, expectedParameterTypes,
-                    expectedReturnType, expectedExceptionParameter, actualMapping, expectedMapping);
+            Set<Method> methods = getMethodsFromClass(iface, name, expectedParameterTypes, expectedReturnType,
+                    expectedExceptionParameter, declaringClass, actualMapping, expectedMapping);
             for (Method method : methods) {
                 if (possibleFallbackMethodNames.contains(method.getName())) {
                     result.add(method);
@@ -180,13 +181,24 @@ final class SecurityActions {
         return result;
     }
 
-    private static Set<Method> getMethodsFromClass(Class<?> guardedMethodDeclaringClass,
-            Class<?> classToSearch, String name, Type[] parameterTypes, Type returnType, boolean exceptionParameter,
+    /**
+     * Returns all methods that:
+     * <ul>
+     * <li>are declared directly on given {@code classToSearch},</li>
+     * <li>have given {@code name},</li>
+     * <li>have matching {@code parameterTypes},</li>
+     * <li>have matching {@code returnType},</li>
+     * <li>have an additional {@code exceptionParameter} if required,</li>
+     * <li>are accessible from given {@code guardedMethodDeclaringClass}.</li>
+     * </ul>
+     */
+    private static Set<Method> getMethodsFromClass(Class<?> classToSearch, String name, Type[] parameterTypes,
+            Type returnType, boolean exceptionParameter, Class<?> guardedMethodDeclaringClass,
             TypeMapping actualMapping, TypeMapping expectedMapping) {
         Set<Method> set = new HashSet<>();
         for (Method method : classToSearch.getDeclaredMethods()) {
-            if (isAccessibleFrom(method, guardedMethodDeclaringClass)
-                    && method.getName().equals(name)
+            if (method.getName().equals(name)
+                    && isAccessibleFrom(method, guardedMethodDeclaringClass)
                     && signaturesMatch(method, parameterTypes, returnType, exceptionParameter,
                             actualMapping, expectedMapping)) {
                 set.add(method);
@@ -211,6 +223,15 @@ final class SecurityActions {
             boolean expectedExceptionParameter, TypeMapping actualMapping, TypeMapping expectedMapping) {
         int expectedParameters = expectedParameterTypes.length;
         if (expectedExceptionParameter) {
+            // need to figure this out _before_ expanding the `expectedParameterTypes` array
+            boolean kotlinSuspendingFunction = KotlinSupport.isSuspendingFunction(expectedParameterTypes);
+            // adjust `expectedParameterTypes` so that there's one more element on the position
+            // where the exception parameter should be, and the value on that position is `null`
+            expectedParameterTypes = Arrays.copyOfRange(expectedParameterTypes, 0, expectedParameters + 1);
+            if (kotlinSuspendingFunction) {
+                expectedParameterTypes[expectedParameters] = expectedParameterTypes[expectedParameters - 1];
+                expectedParameterTypes[expectedParameters - 1] = null;
+            }
             expectedParameters++;
         }
 
@@ -219,18 +240,19 @@ final class SecurityActions {
             return false;
         }
 
-        for (int i = 0; i < expectedParameterTypes.length; i++) {
-            if (!typeMatches(methodParams[i], expectedParameterTypes[i], actualMapping, expectedMapping)) {
-                return false;
-            }
-        }
-
-        if (expectedExceptionParameter) {
-            Type lastParameter = methodParams[methodParams.length - 1];
-            boolean isThrowable = lastParameter instanceof Class
-                    && Throwable.class.isAssignableFrom((Class<?>) lastParameter);
-            if (!isThrowable) {
-                return false;
+        for (int i = 0; i < expectedParameters; i++) {
+            Type methodParam = methodParams[i];
+            Type expectedParamType = expectedParameterTypes[i];
+            if (expectedParamType != null) {
+                if (!typeMatches(methodParam, expectedParamType, actualMapping, expectedMapping)) {
+                    return false;
+                }
+            } else { // exception parameter
+                boolean isThrowable = methodParam instanceof Class
+                        && Throwable.class.isAssignableFrom((Class<?>) methodParam);
+                if (!isThrowable) {
+                    return false;
+                }
             }
         }
 
@@ -273,8 +295,11 @@ final class SecurityActions {
 
     private static boolean parameterizedTypeMatches(ParameterizedType actualType, ParameterizedType expectedType,
             TypeMapping actualMapping, TypeMapping expectedMapping) {
-        return typeArrayMatches(actualType.getActualTypeArguments(), expectedType.getActualTypeArguments(),
+        boolean genericClassMatch = typeMatches(actualType.getRawType(), expectedType.getRawType(),
                 actualMapping, expectedMapping);
+        boolean typeArgumentsMatch = typeArrayMatches(actualType.getActualTypeArguments(),
+                expectedType.getActualTypeArguments(), actualMapping, expectedMapping);
+        return genericClassMatch && typeArgumentsMatch;
     }
 
     private static boolean typeArrayMatches(Type[] actualTypes, Type[] expectedTypes,

--- a/testsuite/basic/src/test/kotlin/io/smallrye/faulttolerance/kotlin/fallback/KotlinFallbackTest.kt
+++ b/testsuite/basic/src/test/kotlin/io/smallrye/faulttolerance/kotlin/fallback/KotlinFallbackTest.kt
@@ -1,0 +1,18 @@
+package io.smallrye.faulttolerance.kotlin.fallback
+
+import io.smallrye.faulttolerance.util.FaultToleranceBasicTest
+import io.smallrye.faulttolerance.util.WithSystemProperty
+import kotlinx.coroutines.runBlocking
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+
+// so that FT methods don't have to be marked @AsynchronousNonBlocking
+@WithSystemProperty(key = "smallrye.faulttolerance.mp-compatibility", value = "false")
+@FaultToleranceBasicTest
+class KotlinFallbackTest {
+    @Test
+    fun test(service: MyService) = runBlocking<Unit> {
+        val result = service.hello("world")
+        assertThat(result).isEqualTo("Hello, WORLD")
+    }
+}

--- a/testsuite/basic/src/test/kotlin/io/smallrye/faulttolerance/kotlin/fallback/MyService.kt
+++ b/testsuite/basic/src/test/kotlin/io/smallrye/faulttolerance/kotlin/fallback/MyService.kt
@@ -1,0 +1,19 @@
+package io.smallrye.faulttolerance.kotlin.fallback
+
+import jakarta.enterprise.context.ApplicationScoped
+import kotlinx.coroutines.delay
+import org.eclipse.microprofile.faulttolerance.Fallback
+
+@ApplicationScoped
+open class MyService {
+    @Fallback(fallbackMethod = "fallback")
+    open suspend fun hello(name: String): String {
+        delay(1000)
+        throw IllegalArgumentException()
+    }
+
+    open suspend fun fallback(name: String, ex: Exception): String {
+        delay(1000)
+        return "Hello, ${name.uppercase()}"
+    }
+}

--- a/testsuite/basic/src/test/kotlin/io/smallrye/faulttolerance/kotlin/fallback/invalid/KotlinInvalidFallbackTest.kt
+++ b/testsuite/basic/src/test/kotlin/io/smallrye/faulttolerance/kotlin/fallback/invalid/KotlinInvalidFallbackTest.kt
@@ -1,0 +1,17 @@
+package io.smallrye.faulttolerance.kotlin.fallback.invalid
+
+import io.smallrye.faulttolerance.util.ExpectedDeploymentException
+import io.smallrye.faulttolerance.util.FaultToleranceBasicTest
+import io.smallrye.faulttolerance.util.WithSystemProperty
+import jakarta.enterprise.inject.spi.DefinitionException
+import org.junit.jupiter.api.Test
+
+// so that FT methods don't have to be marked @AsynchronousNonBlocking
+@WithSystemProperty(key = "smallrye.faulttolerance.mp-compatibility", value = "false")
+@FaultToleranceBasicTest
+@ExpectedDeploymentException(DefinitionException::class)
+class KotlinInvalidFallbackTest {
+    @Test
+    fun test(ignored: MyService) {
+    }
+}

--- a/testsuite/basic/src/test/kotlin/io/smallrye/faulttolerance/kotlin/fallback/invalid/MyService.kt
+++ b/testsuite/basic/src/test/kotlin/io/smallrye/faulttolerance/kotlin/fallback/invalid/MyService.kt
@@ -1,0 +1,17 @@
+package io.smallrye.faulttolerance.kotlin.fallback.invalid
+
+import jakarta.enterprise.context.ApplicationScoped
+import kotlinx.coroutines.delay
+import org.eclipse.microprofile.faulttolerance.Fallback
+
+@ApplicationScoped
+open class MyService {
+    @Fallback(fallbackMethod = "fallback")
+    open suspend fun hello(name: String): String {
+        throw IllegalArgumentException()
+    }
+
+    open suspend fun fallback(name: String, ex: Exception): Int {
+        return 0
+    }
+}

